compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs

```diff
@@ -143,31 +151,220 @@ pub fn expand_deriving_smart_ptr(
 
     // Find the `#[pointee]` parameter and add an `Unsize<__S>` bound to it.
     let mut impl_generics = generics.clone();
+    let pointee_ty_ident = generics.params[pointee_param_idx].ident;
+    let mut self_bounds;
     {
         let p = &mut impl_generics.params[pointee_param_idx];
+        self_bounds = p.bounds.clone();
         let arg = GenericArg::Type(s_ty.clone());
         let unsize = cx.path_all(span, true, path!(span, core::marker::Unsize), vec![arg]);
         p.bounds.push(cx.trait_bound(unsize, false));
         let mut attrs = thin_vec![];
         swap(&mut p.attrs, &mut attrs);
         p.attrs = attrs.into_iter().filter(|attr| !attr.has_name(sym::pointee)).collect();
     }
+    // We should not set default values to constant generic parameters
+    // and commute bounds that indirectly involves `#[pointee]`.
+    for (params, orig_params) in impl_generics.params[pointee_param_idx + 1..]
+        .iter_mut()
+        .zip(&generics.params[pointee_param_idx + 1..])
```

Comment:
Why do we only care about `pointee_param_idx + 1 ..`? Is there some restriction about the placement of `#[pointee]` that this is implicitly relying on?

Reply:
It has to do with those generic types that come after the annotated `#[pointee]` type parameter that may contain bounds "involving" the `#[pointee]` type, such as the `X` in this example.
```rust
#[derive(core::marker::SmartPointer)]
#[repr(transparent)]
pub struct Ptr6<'a, #[pointee] T: ?Sized, X: MyTrait<T>> {
    data: &'a mut T,
    x: core::marker::PhantomData<X>,
}
```

The macro needs to rewrite the bound `X: MyTrait<T>` into `X: MyTrait<__S>` as well.

Reply:
But bounds can reference the `#[pointee]` parameter before the parameter is declared:

```
trait Foo<S>{}

struct Test<S: Foo<T>, #[pointee] T> { /* ... */ }
```

Reply:
^?

Reply:
I always thought this was not possible. I will update this part.

Reply:
Applied. `pointee_param_idx` is skipped because that bound will be written separately.